Create 2021-09-27-solidity-0.8.8-release-announcement.md (#151)
ethereum/solidity-blog: _posts/2021-09-27-solidity-0.8.8-release-announcement.md

---
layout: post
published: true
title: 'Solidity 0.8.8 Release Announcement'
date: '2021-09-27'
author: Solidity Team
category: Releases
---

[Solidity v0.8.8](https://github.com/ethereum/solidity/releases/tag/v0.8.8) introduces user defined
value types as a major feature, some new features, and bugfixes.

## Notable New Features

### User Defined Value Types

A user defined value type allows creating a zero-cost-abstraction over an elementary value type that
also increases type safety and improves readability. Details can be found in the focussed [blog
post](https://deploy-preview-148--solidity-blog.netlify.app/2021/09/13/user-defined-value-types/).

<!-- TODO fix link -->

> ### Inheritance

Author: Is that sub-title expressive enough? Maybe it's too generic?

A function that overrides only a single interface function does not require the override specifier.

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.8;

interface IMinimalERC20 {
     function transfer(address to, uint value) external;
}

contract MinimalERC20 is IMinimalERC20 {
     // Before version 0.8.8, this function declaration required
     // specifying the override keyword, i.e.,
     // `function transfer(address to, uint value) external override ...`
     function transfer(address to, uint value) external {
         // ...
     }
}
```

### `min` and `max` for Enums

The smallest and the largest value of an [enum
type](https://docs.soliditylang.org/en/latest/structure-of-a-contract.html?#enum-types) can now be
accessed using `type(E).min` and `type(E).max` respectively.

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.8;

enum AuctionState {Created, InProgress, Completed}

// returns AuctionState.Created
function min() pure returns (AuctionState) {
    return type(AuctionState).min;
}

// returns AuctionState.Completed
function max() pure returns (AuctionState) {
    return type(AuctionState).max;
}
```

> ### Immutable Variables

Author: Subtitle too generic?

Immutable variables can be read at construction time once they are initialized.

```solidity
// SPDX-License-Identifier: GPL-3.0
pragma solidity ^0.8.8;

interface IMockInterface {
    function getValue() external returns(uint);
}

contract ImmutableExample {
    uint immutable value;

    event setValue(uint);

    constructor(IMockInterface mockContract) {
        // Sets the immutable variable.
        value = mockContract.getValue();
        // Reads the immutable variable.
        emit setValue(value);
    }
}
```

As you probably know, it is also possible to initialize immutable variables
at the point of their declaration, for example `uint immutable deployTime = block.timestamp;`.
Because the order of initialization is not
always obvious, such immutable variables are only considered initialized once
the constructor is executed. This means you can read from such variables
inside the constructor but you cannot use them to initialize other
immutable variables at the point of their declaration.

> ### Import Paths

Author: Subtitle too generic?

We added the new commandline option `--include-path` that hopefully makes it much easier to compile
more complicated projects using multiple libraries. It is available in `solcjs` and `solc` both in legacy
and standard-json mode. You can use it to specify
directories where source files are looked up when `import` statements are resolved. If you use
`npm` to install Solidity packages, they are usually placed in a subdirectory below
`node_modules`. If you use imports of the form `import X from "@packagename/contracts/x.sol";`,
then invoking the compiler as follows should properly resolve all imports without further work
(the base path and the include paths will be added to the "allowed paths", for example):

```bash
solc contract.sol --base-path . --include-path node_modules/
```

The previously introduced option `--base-path` provides a path where the lookup is performed
first. If it fails, the include paths are searched for the file.

It is often a source of frustration that the Solidity compiler stores the full path
to a source file as part of the metadata, which sometimes makes it more difficult to
reproduce a compilation when not using the standard-json mode.

> If you use `--base-path` and `--include-path`, these paths will not be stored in

Reviewer: It reads a bit as if "paths" was referring to file paths rather than paths passed to these options. So someone might think that file paths are not stored in metadata at all in this case. I'd reword it to make it completely clear:
```suggestion
If you use `--base-path` and `--include-path`, then only relative paths to files located inside them will be stored in
```

---
layout: post
published: true
title: 'Solidity 0.8.8 Release Announcement'
date: '2021-09-27'
author: Solidity Team
category: Releases
---

> [Solidity v0.8.7](https://github.com/ethereum/solidity/releases/tag/v0.8.8) introduces .....

Reviewer 2: ```suggestion
[Solidity v0.8.8](https://github.com/ethereum/solidity/releases/tag/v0.8.8) introduces .....
```

> [Solidity v0.8.8](https://github.com/ethereum/solidity/releases/tag/v0.8.8) introduces .....

Reviewer 3: ```suggestion
[Solidity v0.8.8](https://github.com/ethereum/solidity/releases/tag/v0.8.8) introduces user defined value types as a major feature. Details can be found in the [blog post](https://deploy-preview-148--solidity-blog.netlify.app/2021/09/13/user-defined-value-types/). 
<!-- TODO fix link -->
```

immutable variables at the point of their declaration.

### Import Include Paths

We added a new option `--include-path` to `solc` that hopefully makes it much easier to compile

> more complicated projects using multiple libraries. You can use it to specify

Reviewer: Oh, one more thing. We might also say that `--include-path` is meant to be an alternative to import remapping in simple cases (i.e. where they are not used to rename the file itself). For example Brownie currently uses remapping to resolve dependencies and I think that they could switch to this feature instead.

Reviewer 4: Added another paragraph.

immutable variables at the point of their declaration.

### Import Paths

We added a new option `--include-path` to `solc` that hopefully makes it much easier to compile

> more complicated projects using multiple libraries. You can use it to specify

Reviewer: Might be worth clarifying that it affects the default import callback. I.e. all CLI usage, with or without `--standard-json` but not usage via solc-js (unless someone uses the `solcjs` script to invoke the compiler).

Reviewer 4: I considered it, but isn't that use-case a bit too specific and might confuse the "ordinary peopel"? I mean this talk about "options"  - and if you use the binary interface, there is no real place to provide these options, isn't it?

Reviewer: True, but what I'm worried about is that it's not clear that it affects `--standard-json` too. Though maybe it's ok - most people probably are also not aware that with `--standard-json` imports might still get loaded from disk.

directories where source files are looked up when `import` statements are resolved. If you use
`npm` to install Solidity packages, they are usually placed in a subdirectory below
`node_modules`. If you use imports of the form `import X from "@packagename/contracts/x.sol";`,
then invoking the compiler as follows should properly resolve all imports without further work:

```bash
solc contract.sol --base-path . --include-path node_modules/

> ```

Reviewer: Do we want to mention that base path and include paths are automatically added to allowed paths? "without further work" kinda means that but it's not very explicit.

Reviewer 4: yep, will do.
